src/iarc7_motion/state_monitor.py

```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env python
+import actionlib
+import rospy
+import sys
+import threading
+import traceback
+
+from geometry_msgs.msg import TwistStamped
+from geometry_msgs.msg import Twist
+from nav_msgs.msg import Odometry
+
+from iarc7_msgs.msg import TwistStampedArray, OdometryArray
+from iarc7_msgs.msg import FlightControllerStatus
+
+from iarc_tasks.takeoff_task import TakeoffTask
+from iarc_tasks.land_task import LandTask
+from iarc_tasks.test_task import TestTask
+from iarc_tasks.xyztranslation_task import XYZTranslationTask
+from iarc_tasks.track_roomba_task import TrackRoombaTask
+from iarc_tasks.hit_roomba_task import HitRoombaTask
+from iarc_tasks.block_roomba_task import BlockRoombaTask
+from iarc_tasks.hold_position_task import HoldPositionTask
+from iarc_tasks.height_recovery_task import HeightRecoveryTask
+from iarc_tasks.velocity_task import VelocityTask
+
+
+class StateMonitor:
+
+    def __init__(self):
+        # state data
+        self._roombas = None
+        self._obstacles = None
+        self._drone_odometry = None
+        self._arm_status = None
+        
+        # current state of HLM
+        self._initialized = False
+        self._waiting_on_takeoff = True
+        self._waiting_on_recovery = False
+
+        # to keep things thread safe
+        self._lock = threading.RLock()
+        
+        # info needed to do sanity checking and state monitoring
+        self._roomba_status_sub = rospy.Subscriber(
+            'roombas', OdometryArray, 
+            self._receive_roomba_status)
+
+        self._obstacle_sub = rospy.Subscriber(
+            '/obstacles', OdometryArray, 
+            self._receive_obstacle_status)
+
+        self._current_velocity_sub = rospy.Subscriber(
+            '/odometry/filtered', Odometry,
+            self._recieve_drone_odometry)
+
+        self._drone_arm_status = rospy.Subscriber(
+            '/fc_status', FlightControllerStatus, 
+            self._receive_arm_status)
+
+        try:
+            # task timeout time to decelerate
+            self._task_timeout_deceleration_time = rospy.Duration(
+                rospy.get_param('~task_timeout_deceleration_time'))
+            # minimum safe height to manuever at
+            self._MIN_MANEUVER_HEIGHT = rospy.get_param('~min_maneuver_height')
+
+        except KeyError as e:
+            rospy.logerr('Could not lookup a parameter for State Monitor')
+            raise
+
+    # checks task transitions before executing it
+    def check_transition(self, task):
+        with self._lock: 
+            passed = False
+
+            if self._waiting_on_takeoff: 
+                passed = isinstance(task, TakeoffTask)
+            elif isinstance(task, LandTask):
+                self._waiting_on_takeoff = True
```

Comment:
What if landing is aborted?

Reply:
I agree, this needs to be more more robust and a better system for checking.